non-fungible-token/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+
+        self.owner_by_id.insert(token_id, *to);
+
+        let transfer_token = Transfer {
+            from: H256::from_slice(from.as_ref()),
+            to: H256::from_slice(to.as_ref()),
+            token_id,
+        };
+
+        msg::reply(
+            NftEvent::Transfer(transfer_token),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    fn approve(&mut self, owner: &ActorId, spender: &ActorId, token_id: U256) {
+        if spender == &ZERO_ID {
+            panic!("NonFungibleToken: Approval to zero address.");
+        }
+        if spender == owner {
+            panic!("NonFungibleToken: Approval to current owner");
+        }
+        if !self.is_token_owner(token_id, owner) {
+            panic!("NonFungibleToken: is not owner");
+        }
+
+        self.token_approvals.insert(token_id, *spender);
+
+        let approve_token = Approve {
+            owner: H256::from_slice(owner.as_ref()),
+            spender: H256::from_slice(spender.as_ref()),
+            token_id,
+        };
+        msg::reply(
+            NftEvent::Approval(approve_token),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    fn approve_for_all(&mut self, owner: &ActorId, operator: &ActorId, approved: bool) {
+        if operator == &ZERO_ID {
+            panic!("NonFungibleToken: Approval for a zero address");
+        }
+
+        self.operator_approval
+            .entry(*owner)
+            .or_default()
+            .insert(*operator, approved);
+
+        let approve_operator = ApproveForAll {
+            owner: H256::from_slice(owner.as_ref()),
+            operator: H256::from_slice(operator.as_ref()),
+            approved,
+        };
+
+        msg::reply(
+            NftEvent::ApprovalForAll(approve_operator),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+}
+
+impl NonFungibleToken {
+    pub fn is_token_owner(&self, token_id: U256, account: &ActorId) -> bool {
+        account == self.owner_by_id.get(&token_id).unwrap_or(&ZERO_ID)
+    }
+
+    pub fn authorized_actor(&self, token_id: U256, account: &ActorId) -> AuthAccount {
+        let owner = self.owner_by_id.get(&token_id).unwrap_or(&ZERO_ID);
+        if owner == account {
+            return AuthAccount::Owner;
+        }
+        if self.token_approvals.get(&token_id).unwrap_or(&ZERO_ID) == account {
+            return AuthAccount::ApprovedActor;
+        }
+        if *self
+            .operator_approval
+            .get(owner)
+            .unwrap_or(&BTreeMap::<ActorId, bool>::default())
+            .get(account)
+            .unwrap_or(&false)
+        {
+            return AuthAccount::Operator;
+        }
+        AuthAccount::None
+    }
+
+    pub fn exists(&self, token_id: U256) -> bool {
+        self.owner_by_id.contains_key(&token_id)
+    }
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub struct Approve {
```

Comment:
What is this used for? In #10 it is redundantly imported and doesn't hold any functionality as far as I understand.

Reply:
The structs `Approve`, `ApproveForAll`, `Transfer` are used in msg::reply(...) (enum `Events`)